### PR TITLE
nixos/krb5: complete rewrite

### PR DESCRIPTION
**Pull request by eqyiel**
###### Motivation for this change

The `krb5` service was a bit lacking.

Addresses NixOS/nixpkgs#11268, partially addresses NixOS/nixpkgs#29623.

I've made an effort to accommodate people who are using the old configuration keys.

Given the state that it was in though, I wonder anyone even used this module.

Also I'm not sure if this is an abuse of the testing framework.  Snapshot testing is something I do a lot of in web development so it makes sense to me to try it here.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
Tests looks ok, but maybe launching a vm is not necessary. @Profpatsch do you have a better idea here?

File: nixos/modules/config/krb5/default.nix [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        '';
+      };
+
+      defaultRealm = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "ATHENA.MIT.EDU";
+        description = ''
+          DEPRECATED, please use
+          <literal>krb5.libdefaults.default_realm</literal>.
```
Comment (Mic92) [inline nixos/modules/config/krb5/default.nix]: Are you aware of `modules/rename.nix`?

Comment (eqyiel) [reply nixos/modules/config/krb5/default.nix]: TIL, thanks!

Comment (eqyiel) [reply nixos/modules/config/krb5/default.nix]: As far as I can tell, this only forwards definitions for deprecated options that have a one-to-one mapping to non-deprecated options (or just prints a warning saying what to do).  Do you think it's worth it to use `mkRenamedOptionModule` for the only deprecated option that does have a one-to-one relationship with a non-deprecated option (the one you commented on), or to go all in on `mkRenamedOptionModule` and say that those deprecated options are no longer supported at all?

Comment (Mic92) [reply nixos/modules/config/krb5/default.nix]: Yeah probably not that useful here.

Comment (eqyiel) [reply nixos/modules/config/krb5/default.nix]: That's how I feel about it too, but not strongly - I would be happy to change it if that is the preferred way of doing things!

Comment (Mic92) [reply nixos/modules/config/krb5/default.nix]: I added a deprecation date instead

**Comment by Mic92** [comment]
I successfully transformed my old krb5 configuration to the new module syntax.

**Comment by Mic92** [comment]
Nice work. I also gave up on the old module back then.

**Comment by eqyiel** [comment]
Thanks!!

**Comment by Profpatsch** [comment]
> Tests looks ok, but maybe launching a vm is not necessary. @Profpatsch do you have a better idea here?

For service modules it’s kind of hard to test the correct functionality without a NixOS VM. :)
If you want to test just the executable though, that could be done. Of course one can also mix.

**Pull request merged**